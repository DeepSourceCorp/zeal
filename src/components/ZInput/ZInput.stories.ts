import '../../assets/css/tailwind.css'
import '../../assets/css/typography.css'
import '../../assets/css/layout.css'
import ZInput from './ZInput.vue'
import ZInputGroup from '@/components/ZInputGroup/ZInputGroup.vue'
import ZInputAddon from '@/components/ZInputAddon/ZInputAddon.vue'
import ZIcon from '@/components/ZIcon/ZIcon.vue'
import ZButton from '@/components/ZButton/ZButton.vue'
import Vue from 'vue'

export default {
  title: 'Input',
  component: ZInput,
  excludeStories: /.*Data$/
}

export const InputWithCustomPlaceholder = () => ({
  components: { ZInput },
  data() {
    return {
      name: ''
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
            <z-input v-model="name"
                    placeholder="captainmarvel@marvel.com"></z-input>
        </div>
    </div>`
})

export const DebounceInput = () =>
  Vue.extend({
    components: { ZInput },
    data() {
      return {
        name: '',
        delayname: '',
        debouncedValue: '',
        delayDebouncedValue: ''
      }
    },
    methods: {
      debounce(val: string): void {
        this.debouncedValue = val
      },
      delayDebounce(val: string): void {
        this.delayDebouncedValue = val
      }
    },
    template: `<div class='padded-container space-y-6'>
        <div class="input-container space-y-2">
            <z-input @debounceInput="debounce" v-model="name" :showBorder="false" backgroundColor="ink-200"></z-input>
            <p class="text-vanilla-100 text-sm">Debounced Value: {{debouncedValue}}</p>
        </div>
        <div class="input-container space-y-2">
            <z-input :debounceDelay="1200" @debounceInput="delayDebounce" v-model="delayname" :showBorder="false" backgroundColor="ink-200"></z-input>
            <p class="text-vanilla-100 text-sm">Delayed Debounced Value: {{delayDebouncedValue}}</p>
        </div>
    </div>`
  })

export const DisabledInput = () => ({
  components: { ZInput },
  data() {
    return {
      name: 'Hello World'
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
            <z-input v-model="name" 
                    :disabled="true"></z-input>
        </div>
    </div>`
})

export const InputWithDifferentSizes = () => ({
  components: { ZInput, ZButton },
  data() {
    return {
      name: ''
    }
  },
  template: `<div class='padded-container'>
        <div class="max-w-md space-y-2">
          <div class="flex items-center space-x-2">
            <z-input v-model="name" size="x-small"></z-input>
          </div>  
          <div class="flex items-center space-x-2">
            <z-input v-model="name" size="small"></z-input>
          </div>  
          <div class="flex items-center space-x-2">
            <z-input v-model="name" size="medium"></z-input>
          </div>  
          <div class="flex items-center space-x-2">
            <z-input v-model="name" size="large"></z-input>
          </div>  
          <div class="flex items-center space-x-2">
            <z-input v-model="name" size="xlarge"></z-input>
          </div>  
        </div>
    </div>`
})

export const InputBorderAndBackground = () => ({
  components: { ZInput },
  data() {
    return {
      name: ''
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
            <z-input v-model="name" :showBorder="false" backgroundColor="ink-200"></z-input>
        </div>
    </div>`
})

export const StandAloneInput = () => ({
  components: { ZInput },
  data() {
    return {
      name: ''
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
            <z-input v-model="name"></z-input>
        </div>
    </div>`
})

export const InputWithLeftIcon = () => ({
  components: { ZInput, ZInputGroup, ZIcon },
  data() {
    return {
      name: ''
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
          <z-input v-model="name">
            <template slot="left">
              <z-icon icon="search" size="small"></z-icon>
            </template>
          </z-input>
        </div>
    </div>`
})

export const InputWithRightIcon = () => ({
  components: { ZInput, ZInputGroup, ZIcon },
  data() {
    return {
      name: ''
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
          <z-input v-model="name">
            <template slot="right">
              <z-icon icon="search" size="small"></z-icon>
            </template>
          </z-input>
        </div>
    </div>`
})

export const InputWithIconsOnEitherSides = () => ({
  components: { ZInput, ZInputGroup, ZIcon },
  data() {
    return {
      name: ''
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
          <z-input v-model="name">
            <template slot="left">
              <z-icon icon="search" size="small"></z-icon>
            </template>
            <template slot="right">
              <z-icon icon="check" size="small"></z-icon>
            </template>
          </z-input>
        </div>
    </div>`
})

export const InputWithLeftButton = () => ({
  components: { ZInput, ZInputGroup, ZButton },
  data() {
    return {
      name: ''
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
          <z-input v-model="name" size="medium">
            <template slot="left">
              <z-button buttonType="primary" size="small" icon="chevron-left" iconSize="medium" iconColor="ink-400"></z-button>
            </template>
          </z-input>
        </div>
    </div>`
})

export const InputWithRightButton = () => ({
  components: { ZInput, ZInputGroup, ZButton },
  data() {
    return {
      name: ''
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
          <z-input v-model="name" size="medium">
            <template slot="right">
              <z-button buttonType="primary" size="small" icon="chevron-right" iconSize="medium" iconColor="ink-400"></z-button>
            </template>
          </z-input>
        </div>
    </div>`
})

export const InputWithClearables = () => ({
  components: { ZInput, ZInputGroup, ZIcon },
  data() {
    return {
      name: 'Hello World'
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
          <z-input v-model="name" :clearable="true">
          </z-input>
        </div>
    </div>`
})

export const InputWithLeftAddon = () => ({
  components: { ZInput, ZInputGroup, ZInputAddon, ZIcon },
  data() {
    return {
      name: 'Hello World'
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
          <z-input-group>
            <z-input-addon class="bg-juniper px-2 rounded-sm -mr-0.5">
              www.
            </z-input-addon>
            <z-input v-model="name"></z-input>
          </z-input-group>
        </div>
    </div>`
})

export const InputWithRightAddon = () => ({
  components: { ZInput, ZInputGroup, ZInputAddon, ZIcon },
  data() {
    return {
      name: 'Hello World'
    }
  },
  template: `<div class='padded-container'>
        <div class="input-container">
          <z-input-group>
            <z-input v-model="name"></z-input>
            <z-input-addon class="bg-juniper px-2 rounded-sm -ml-0.5">
              www.
            </z-input-addon>
          </z-input-group>
        </div>
    </div>`
})
