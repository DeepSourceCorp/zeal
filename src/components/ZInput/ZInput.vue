<template>
  <div
    class="flex items-center w-full rounded-sm outline-none space-x-2 p-1"
    :class="[
      borderStyles,
      `bg-${backgroundColor}`,
      {
        'text-slate cursor-not-allowed': disabled,
        'text-vanilla-300': !disabled,
        'h-6 text-xs space-x-1 leading-loose': size === 'x-small',
        'h-8 text-xs space-x-1 leading-loose': size === 'small',
        'h-10 text-sm space-x-2 leading-8': size === 'medium',
        'h-12 text-base space-x-2.5 leading-9': size === 'large',
        'h-13 text-lg space-x-3 leading-9': ['x-large', 'xlarge'].includes(size),
        'py-0.5 px-1.5': size === 'x-small' && !padding,
        'py-1 px-2': size === 'small' && !padding,
        'py-2 px-3': size === 'medium' && !padding,
        'py-2.5 px-4': size === 'large' && !padding,
        'py-3 px-4': ['x-large', 'xlarge'].includes(size) && !padding,
        'pr-1': $slots.right && !$slots.left && !padding,
        'pl-1': $slots.left && !$slots.right && !padding,
        [padding]: padding
      }
    ]"
  >
    <!-- Any icon/content to the left renders here -->
    <slot name="left"></slot>
    <input
      v-debounce:[debounceDelay]="updateDebounce"
      class="overflow-ellipsis overflow-hidden w-full caret-juniper flex flex-grow outline-none bg-transparent"
      :class="{
        'cursor-not-allowed': disabled
      }"
      :aria-label="label"
      :value="name"
      :type="type"
      :max="max"
      :maxlength="maxLength"
      :min="min"
      :minlength="minLength"
      :multiple="multiple"
      :required="required"
      :placeholder="placeholder"
      :disabled="disabled"
      :autocomplete="autocomplete"
      @input="updateSelf($event.target.value)"
      @blur="this.blurHandler"
      @focus="this.focusHandler"
      @keydown="this.keydownHandler"
      @keyup="this.keyupHandler"
      @invalid="this.invalidHandler"
    />
    <!-- Any icon/content to the right renders here -->
    <slot name="right">
      <z-icon v-if="clearable" icon="x" size="small" class="cursor-pointer" @click.stop="updateSelf('')"> </z-icon>
    </slot>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import ZIcon from '../ZIcon/ZIcon.vue'
import debounceDirective from '@/directives/debounce'

export default Vue.extend({
  name: 'ZInput',
  components: {
    ZIcon
  },
  directives: {
    debounce: debounceDirective
  },
  props: {
    name: {
      default: '',
      type: String
    },
    label: {
      default: '',
      type: String
    },
    type: {
      default: 'text',
      type: String
    },
    padding: {
      type: String,
      default: undefined
    },
    max: {
      type: [String, Number],
      default: undefined
    },
    min: {
      type: [String, Number],
      default: undefined
    },
    placeholder: {
      default: 'Enter a value',
      type: String
    },
    disabled: {
      default: false,
      type: Boolean
    },
    debounceDelay: {
      default: 350,
      type: Number
    },
    icon: {
      type: String,
      default: undefined
    },
    clearable: {
      type: Boolean
    },
    autocomplete: {
      type: String,
      default: undefined
    },
    size: {
      default: 'medium',
      type: String,
      validator(val) {
        return ['x-small', 'small', 'medium', 'large', 'x-large'].includes(val)
      }
    },
    backgroundColor: {
      type: String,
      default: 'ink-400'
    },
    showBorder: {
      type: Boolean,
      default: true
    },
    isInvalid: {
      type: Boolean,
      default: false
    },
    errorBorderColor: {
      type: String,
      default: 'cherry'
    },
    maxLength: {
      type: [String, Number],
      default: undefined
    },
    minLength: {
      type: [String, Number],
      default: undefined
    },
    multiple: {
      type: Boolean,
      default: false
    },
    required: {
      type: Boolean,
      default: false
    }
  },
  model: {
    prop: 'name',
    event: 'input'
  },
  data() {
    return {
      invalidState: false
    }
  },
  computed: {
    borderStyles(): string {
      if (this.isInvalid || this.invalidState) {
        return `border border-cherry`
      }

      if (this.showBorder) {
        return `focus-within:border-vanilla-400 border border-ink-100`
      }

      return ''
    }
  },
  methods: {
    updateSelf(newValue: string): void {
      this.$emit('input', newValue)
    },
    updateDebounce(value: unknown): void {
      this.$emit('debounceInput', value)
    },
    blurHandler(e: FocusEvent) {
      this.$emit('blur', e)
      const eventTarget = e.target as HTMLInputElement
      if (!eventTarget.checkValidity()) {
        this.invalidSate = true
      } else {
        this.invalidSate = false
      }
    },
    focusHandler(e: FocusEvent) {
      this.$emit('focus', e)
    },
    keyupHandler(e: KeyboardEvent) {
      this.$emit('keyup', e)
    },
    keydownHandler(e: KeyboardEvent) {
      this.$emit('keydown', e)
    },
    invalidHandler(e: Event) {
      this.invalidState = true
      this.$emit('invalid', e)
    }
  }
})
</script>
