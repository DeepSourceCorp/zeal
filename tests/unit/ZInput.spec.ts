import ZInput from '../../src/components/ZInput'
import Vue from 'vue'

import { mount, Wrapper } from '@vue/test-utils'

describe('Input Component', () => {
  let mountFn: (options?: Record<string, unknown>) => Wrapper<Vue>
  let localVue: typeof Vue

  beforeEach(() => {
    mountFn = (options = {}) => {
      return mount(ZInput, {
        localVue,
        ...options
      })
    }
  })

  it('renders the default input component', () => {
    const wrapper = mountFn()

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll('input').length).toEqual(1)
  })
  it('renders the input with custom placeholder', () => {
    const wrapper = mountFn({
      propsData: {
        placeholder: 'captain@marvel.com'
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('input').attributes('placeholder')).toMatch('captain@marvel.com')
  })
  it('renders a disabled input', () => {
    const wrapper = mountFn({
      propsData: {
        disabled: true
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('input').attributes('disabled')).toMatch('disabled')
  })
  it('renders an input with clearable', () => {
    const wrapper = mountFn({
      propsData: {
        clearable: true
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll('.z-icon--x').length).toEqual(1)
  })
})
