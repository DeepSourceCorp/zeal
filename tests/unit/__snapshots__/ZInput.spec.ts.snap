// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component emits debounce changes 1`] = `
<div class="flex items-center w-full rounded-sm outline-none space-x-2 p-1 focus-within:border-vanilla-400 border border-solid border-ink-100 bg-ink-400 text-vanilla-300 h-10 py-2 text-base space-x-2 leading-8 py-2 px-4"> <input aria-label="" type="text" placeholder="captain@marvel.com" class="w-full caret-juniper flex flex-grow outline-none bg-transparent pl-0.5">
  <!---->
</div>
`;

exports[`Input Component renders a disabled input 1`] = `
<div class="flex items-center w-full rounded-sm outline-none space-x-2 p-1 focus-within:border-vanilla-400 border border-solid border-ink-100 bg-ink-400 text-slate cursor-not-allowed h-10 py-2 text-base space-x-2 leading-8 py-2 px-4"> <input aria-label="" type="text" placeholder="Enter a value" disabled="disabled" class="w-full caret-juniper flex flex-grow outline-none bg-transparent pl-0.5 cursor-not-allowed">
  <!---->
</div>
`;

exports[`Input Component renders an input with clearable 1`] = `
<div class="flex items-center w-full rounded-sm outline-none space-x-2 p-1 focus-within:border-vanilla-400 border border-solid border-ink-100 bg-ink-400 text-vanilla-300 h-10 py-2 text-base space-x-2 leading-8 py-2 px-4"> <input aria-label="" type="text" placeholder="Enter a value" class="w-full caret-juniper flex flex-grow outline-none bg-transparent pl-0.5"> <svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="cursor-pointer z-icon w-4 h-4 z-icon--x text-vanilla-400">
    <line x1="18" y1="6" x2="6" y2="18"></line>
    <line x1="6" y1="6" x2="18" y2="18"></line>
  </svg></div>
`;

exports[`Input Component renders the default input component 1`] = `
<div class="flex items-center w-full rounded-sm outline-none space-x-2 p-1 focus-within:border-vanilla-400 border border-solid border-ink-100 bg-ink-400 text-vanilla-300 h-10 py-2 text-base space-x-2 leading-8 py-2 px-4"> <input aria-label="" type="text" placeholder="Enter a value" class="w-full caret-juniper flex flex-grow outline-none bg-transparent pl-0.5">
  <!---->
</div>
`;

exports[`Input Component renders the input with custom placeholder 1`] = `
<div class="flex items-center w-full rounded-sm outline-none space-x-2 p-1 focus-within:border-vanilla-400 border border-solid border-ink-100 bg-ink-400 text-vanilla-300 h-10 py-2 text-base space-x-2 leading-8 py-2 px-4"> <input aria-label="" type="text" placeholder="captain@marvel.com" class="w-full caret-juniper flex flex-grow outline-none bg-transparent pl-0.5">
  <!---->
</div>
`;
